test/BulletDynamics/test_deterministic_save_and_restore.cpp

```diff
@@ -0,0 +1,165 @@
+#include "LinearMath/btVector3.h"
+
+#include <gtest/gtest.h>
+
+static btMultiBodyDynamicsWorld *create_btMultiBodyDynamicsWorld()
+{
+	btDefaultCollisionConfiguration * collisionConfiguration = new btDefaultCollisionConfiguration();
+
+	btCollisionDispatcher * dispatcher = new btCollisionDispatcher(collisionConfiguration);
+
+	btDbvtBroadphase * broadphase = new btDbvtBroadphase();
+
+	btMultiBodyConstraintSolver* solver = new btMultiBodyConstraintSolver();
+
+	btMultiBodyDynamicsWorld * world = new btMultiBodyDynamicsWorld(dispatcher, broadphase, solver, collisionConfiguration);
+
+	world->setGravity(btVector3(0, -10, 0));
+
+	return world;
+
+}
+
+btRigidBody* CreateBox(btVector3 position, btVector3 velocity, btVector3 inertia, btScalar mass = 1.f)
+{
+	btBoxShape* colShape = new btBoxShape(btVector3(.1, .1, .1));
+	btTransform startTransform;
+	startTransform.setIdentity();
+	startTransform.setOrigin(position);
+	colShape->calculateLocalInertia(mass, inertia);
+	btRigidBody* box = new btRigidBody(mass, 0, colShape, inertia);
+	box->setWorldTransform(startTransform);
+	box->setUserIndex(-1);
+	box->setLinearVelocity(velocity);
+	return box;
+}
+
+GTEST_TEST(BulletDynamics, DeterministicSaveRestore)
+{
+
+	static char filename[] = "test_serialize.bullet";
+	btRigidBody* box_before[3];
+	btCollisionObject* box_after[3];
+	btMultiBodyDynamicsWorld *initial_world = create_btMultiBodyDynamicsWorld();
+
+	// create ground plane and box shapes
+	// in test three boxes are created
+
+	btBoxShape* groundShape = new btBoxShape(btVector3(btScalar(50.), btScalar(50.), btScalar(50.)));
+	{
+		btTransform groundTransform;
+		groundTransform.setIdentity();
+		groundTransform.setOrigin(btVector3(0, -50, 0));
+		btScalar mass(0.);
+		btVector3 localInertia(0, 0, 0);
+		groundShape->calculateLocalInertia(mass, localInertia);
+		btRigidBody* body = new btRigidBody(mass, 0, groundShape, localInertia);
+		body->setWorldTransform(groundTransform);
+		body->setUserIndex(-1);
+		initial_world->addRigidBody(body);
+	}
+	{
+		btVector3 localInertia(0, 0, 0);
+
+	//set first position of boxes and linear velocity
+
+		box_before[0] = CreateBox(btVector3(2,10,0), btVector3(-1,0,1), localInertia);
+		initial_world->addRigidBody(box_before[0]);
+		box_before[1] = CreateBox(btVector3(2,10,2), btVector3(-1,0,-1), localInertia);
+		initial_world->addRigidBody(box_before[1]);
+		box_before[2] = CreateBox(btVector3(0,10,1), btVector3(1,0,0), localInertia);
+		initial_world->addRigidBody(box_before[2]);
+	}
+
+	// step simulation until collision
+
+	const double delta_t = 0.01;
+	int time = 0;
+	while (box_before[0]->getCenterOfMassPosition().getY() == box_before[1]->getCenterOfMassPosition().getY() &&
+			box_before[1]->getCenterOfMassPosition().getY() == box_before[2]->getCenterOfMassPosition().getY())
+	{
+		time++;
+		initial_world->stepSimulation(delta_t);
+	}
+
+	// serialize
+
+	if (initial_world->getDispatcher()->getNumManifolds() < 2)
```

Comment:
This should be the condition of the while statement, so that the simulation runs until it is true  (plus a couple more frames after that). Otherwise the test will fail without having actually tested anything...